src/pudl/output/ferc1.py

```diff
@@ -1862,87 +2062,84 @@ def forest(self: Self) -> nx.DiGraph:
-            child = [
-                n
-                for n in forest.successors(node)
-                if not n.xbrl_factoid.endswith("_correction")
-            ]
-            correction = [
-                n
-                for n in forest.successors(node)
-                if n.xbrl_factoid.endswith("_correction")
-            ]
-            assert len(child) == 1
-            logger.debug(
-                f"Replacing passthrough node {node} with edge from "
-                f"{parent[0]} to {child[0]}"
-            )
-            forest.remove_nodes_from(correction + [node])
-            forest.add_edge(parent[0], child[0])
-
-        if not nx.is_forest(forest):
-            logger.error(
-                "Calculations in Exploded Metadata can not be represented as a forest!"
-            )
-        connected_components = list(nx.connected_components(forest.to_undirected()))
-        logger.debug(
-            f"Calculation forest contains {len(connected_components)} connected components."
-        )
-
-        # Remove any node that:
-        # - ONLY has stepchildren.
-        # - AND has utility_type total
-        for node in self.stepparents(forest):
+        forest = deepcopy(self.seeded_digraph)
+        # Remove any node that ONLY has stepchildren.
+        pure_stepparents = []
+        stepparents = sorted(self.stepparents(forest))
+        logger.info(f"Investigating {len(stepparents)=}")
+        for node in stepparents:
             children = set(forest.successors(node))
             stepchildren = set(self.stepchildren(forest)).intersection(children)
-            if (
-                (children == stepchildren)
-                & (len(children) > 0)
-                & (node.utility_type == "total")
-            ):
+            if (children == stepchildren) & (len(children) > 0):
+                pure_stepparents.append(node)
                 forest.remove_node(node)
+        logger.info(f"Removed {len(pure_stepparents)} redundant/stepparent nodes.")
+        logger.debug(f"Removed redunant/stepparent nodes: {sorted(pure_stepparents)}")
 
-        # Prune any newly disconnected nodes resulting from the above removal of
-        # pure stepparents. We expect the set of newly disconnected nodes to be empty.
+        # Removing pure stepparents should NEVER disconnect nodes from the forest.
+        # Defensive check to ensure that this is actually true
         nodes_before_pruning = forest.nodes
         forest = self.prune_unrooted(forest)
         nodes_after_pruning = forest.nodes
-
         if pruned_nodes := set(nodes_before_pruning).difference(nodes_after_pruning):
             raise AssertionError(f"Unexpectedly pruned stepchildren: {pruned_nodes=}")
 
-        forest = self.set_forest_attributes(
-            forest,
-            exploded_meta=self.exploded_meta,
-            exploded_calcs=self.exploded_calcs,
-            tags=self.tags,
-        )
+        # HACK alter.
```

Comment:
What does this mean?

Reply:
@cmgosnell made this note and I've also been confused by it.